agent/consul/gateways/bind.go

```diff
@@ -0,0 +1,98 @@
+package gateways
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul/agent/configentry"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// referenceSet stores an O(1) accessible set of ResourceReference objects.
+type referenceSet = map[structs.ResourceReference]any
+
+// gatewayRefs maps a gateway kind/name to a set of resource references.
+type gatewayRefs = map[configentry.KindName][]structs.ResourceReference
+
+// BindRoutesToGateways takes a slice of bound API gateways and a slice of routes.
+// It iterates over the parent references for each route. These parents are gateways the
+// route should be bound to. If the parent matches a bound gateway, the route is bound to
+// the gateway. Otherwise, the route is unbound from the gateway if it was bound.
+//
+// The function returns a list of references to the modified BoundAPIGatewayConfigEntry objects,
+// a map of resource references to errors that occurred when they were attempted to be
+// bound to a gateway, and an error if the overall process was unsucessful.
```

Comment:
#nit I would probably just comment that this is a convenience function for calling `BindRouteToGateways` with multiple routes; otherwise, I typically see that anyone modifying the other function will forget to update this doc.

Reply:
One other thought: is there value in having two separate functions - `BindRoute` and `BindRoutes` - vs. just one function that accepts any number of routes?
```go
func BindRoutesToGateways([]*structs.BoundAPIGatewayConfigEntry, routes ...structs.BoundRoute)
```

Reply:
This was an AWESOME suggestion. It is now variadic